src/ad-targeting-youtube.ts

```diff
@@ -0,0 +1,109 @@
+import { onConsentChange } from '@guardian/consent-management-platform';
+import type { ConsentState } from '@guardian/consent-management-platform/dist/types';
+import { getCookie } from '@guardian/libs';
+import { canUseDom } from './lib/can-use-dom';
+import { constructQuery } from './lib/construct-query';
+import { getPermutivePFPSegments } from './permutive';
+import type {
+	AdsConfig,
+	AdsConfigBasic,
+	AdsConfigCCPAorAus,
+	AdsConfigDisabled,
+	AdsConfigTCFV2,
+} from './types';
+
+type MaybeArray<T> = T[] | T;
+
+type CustomParams = Record<string, MaybeArray<string | number | boolean>>;
+
+const buildCustomParamsFromCookies = (): CustomParams =>
+	canUseDom()
+		? {
+				permutive: getPermutivePFPSegments(),
+				si: getCookie({ name: 'GU_U' }) ? 't' : 'f',
+		  }
+		: {};
+
+const buildAdsConfig = (
+	cmpConsent: ConsentState,
+	adUnit: string,
+	customParams: CustomParams,
+): AdsConfig => {
+	const mergedCustomParams = {
+		...customParams,
+		...buildCustomParamsFromCookies(),
+	};
+
+	const defaultAdsConfig: AdsConfigBasic = {
+		adTagParameters: {
+			iu: adUnit,
+			// TODO: Why are we double encoding? Following Frontend process for now
+			cust_params: encodeURIComponent(constructQuery(mergedCustomParams)),
+		},
+	};
+
+	if (cmpConsent.ccpa) {
+		const canTarget = !cmpConsent.ccpa.doNotSell;
+		return {
+			...defaultAdsConfig,
+			restrictedDataProcessor: !canTarget,
+		} as AdsConfigCCPAorAus;
+	}
+
+	if (cmpConsent.aus) {
+		const canTarget = cmpConsent.aus.personalisedAdvertising;
+		return {
+			...defaultAdsConfig,
+			restrictedDataProcessor: !canTarget,
+		} as AdsConfigCCPAorAus;
+	}
+
+	const tcfData = cmpConsent.tcfv2;
+	// ConsentState type allows for an undefined tcfv2
+	if (tcfData) {
+		const canTarget = Object.values(tcfData.consents).every(Boolean);
```

Comment:
For consistency, I would follow the same pattern as above:

```suggestion
	if (cmpConsent.tcfv2) {
		const tcfData = cmpConsent.tcfv2;
		const canTarget = Object.values(tcfData.consents).every(Boolean);
```